common/tokenbucket/tb_test.go

```diff
@@ -167,62 +140,159 @@ func (s *TokenBucketSuite) TestPriorityLowRpsEnforced() {
-		s.False(ok2)
+		assert.False(t, ok2)
 		ok2, _ = tb.GetToken(1, 1)
-		s.True(ok2)
+		assert.True(t, ok2)
 	}
 }
 
-func (s *TokenBucketSuite) TestFullPriorityTokenBucket() {
-	ts := &mockTimeSource{currTime: time.Now()}
+func TestFullPriorityTokenBucket(t *testing.T) {
+	ts := clock.NewMockedTimeSource()
 	tb := NewFullPriorityTokenBucket(2, 100, ts)
 
 	ok2, _ := tb.GetToken(1, 10)
-	s.True(ok2)
+	assert.True(t, ok2)
 
 	for i := 0; i < 2; i++ {
 		ok2, _ := tb.GetToken(1, 1)
-		s.False(ok2)
+		assert.False(t, ok2)
 		ok, _ := tb.GetToken(0, 10)
-		s.True(ok)
-		ts.advance(time.Millisecond * 101)
+		assert.True(t, ok)
+		ts.Advance(time.Millisecond * 101)
 	}
 
 	ok2, _ = tb.GetToken(1, 1)
-	s.False(ok2)
-	ts.advance(time.Millisecond * 101)
+	assert.False(t, ok2)
+	ts.Advance(time.Millisecond * 101)
 	ok2, _ = tb.GetToken(1, 5)
-	s.True(ok2)
-	ts.advance(time.Millisecond * 101)
+	assert.True(t, ok2)
+	ts.Advance(time.Millisecond * 101)
 	ok2, _ = tb.GetToken(1, 15)
-	s.False(ok2)
+	assert.False(t, ok2)
 	ok2, _ = tb.GetToken(1, 10)
-	s.True(ok2)
+	assert.True(t, ok2)
 	ok, _ := tb.GetToken(0, 10)
-	s.True(ok)
+	assert.True(t, ok)
+}
+
+func TestTokenBucketConsume(t *testing.T) {
+	t.Run("consume_wait", func(t *testing.T) {
+		// make sure we don't deadlock inside the test.
+		go panicOnTimeout(time.Minute)
+
+		ts := clock.NewMockedTimeSource()
+
+		// I provide 10 rps, which means I can consume 10 tokens per second.
+		// tokenBucketImpl will fill 1 token and then refill it every 100 milliseconds.
+		tb := New(10, ts)
+
+		// I consume 1 token, so the next Consume call will block until refill.
+		ok, _ := tb.TryConsume(1)
+		assert.True(t, ok)
+
+		consumeFinished := make(chan struct{})
+		go func() {
+			success := tb.Consume(1, 2*refillRate)
+			assert.True(t, success, "Consume must acquire token successfully")
+			close(consumeFinished)
+		}()
+
+		// I need to make sure that goroutine with Consume call is blocked on time.Sleep.
+		// BlockUntil awaits for at least 1 goroutine to be blocked on time.Sleep/timer/ticker.
+		ts.BlockUntil(1)
+		// Checking that consume is still blocked.
+		select {
+		case <-consumeFinished:
+			assert.Fail(t, "Consume returned before refill")
+		default:
+		}
+		// tb has internal retries to acquire tokens which happens at most backoffInterval (10 * time.Millisecond).
+		// I advance time once, so the next iteration is still blocked.
+		ts.Advance(backoffInterval + 1)
+		// wait until Consume did one iteration and stopped on time.Sleep
+		ts.BlockUntil(1)
+		// Checking that consume is still blocked.
+		select {
+		case <-consumeFinished:
+			assert.Fail(t, "Consume returned before refill")
+		default:
+		}
+		// Advance time to the next refill time.
+		ts.Advance(refillRate - backoffInterval + 1)
+		// Advance should unblock Consume and it should return afterwards.
+		// I don't provide a timeout, since I have a goroutine to panic on timeout.
+		select {
+		case <-consumeFinished:
+		}
```

Comment:
```suggestion
		<-consumeFinished
```
may also be worth mentioning that the goroutine-panic is there to dump the goroutines, which includes the test (waiting here) and the goroutine trying to consume with a timeout.  (because a timeout would have to manually dump goroutines, and that seems like more work than it's worth)